### PR TITLE
Add `_NAGLToolkitWrapper`

### PR DESCRIPTION
**Pull request by mattwthompson**
- [x] Include in (a small portion of) CI matrix
- [x] Warn if conformer-specific arguments are passed through
- [ ] Any other methods that need to be implemented?
- [ ] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/main/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/main/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/main/docs/releasehistory.rst)

### CODE REVIEW

File: openff/toolkit/utils/_nagl_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        if strict_n_conformers:
+            raise Exception
+
+        if partial_charge_method == "_nagl_am1bccelf10":
+            # I can't yet tell from the source code where in NAGL a high-level
+            # `def charge(Molecule, ...):` function is implemented, but I'm sure
+            # it's there somewhere
+            from openff.nagl import charge
+
+            charge(molecule)
```
[the remark this answers is not in the document]
Comment (mattwthompson) [reply openff/toolkit/utils/_nagl_wrapper.py:76]: @lilyminium is there an simple, high-level API point in NAGL for applying partial charges to a single `Molecule`?

[the remark this answers is not in the document]
Comment (lilyminium) [reply openff/toolkit/utils/_nagl_wrapper.py:76]: This is in progress as part of productionalising nagl, as applying charges depends on the structure of the model loaded. The below code is what I use, though, and the final NAGL function will probably look similar (`model.compute_property`); I imagined which particular `model_file` to use would live inside the toolkitwrapper.

```python
def assign_charges_with_nagl_model(offmol: Molecule, model_file: str = None) -> np.ndarray:
    import torch

    model = torch.load(model_file)
    model.eval()

    with capture_toolkit_warnings():
        charges = model.compute_property(offmol).detach().cpu().numpy()
        charges = charges.flatten().astype(float)
        charges = np.asarray(charges)

    return charges
```

[the remark this answers is not in the document]
Comment (mattwthompson) [reply openff/toolkit/utils/_nagl_wrapper.py]: With https://github.com/openforcefield/openff-nagl/pull/26 coming along there is not a pretty easy way to access this functionality

File: openff/toolkit/utils/_nagl_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        if strict_n_conformers:
+            warnings.warn(
+                "`_NAGLToolkitWrapper.assign_partial_charges` was passed optional orgument "
+                "`strict_n_conformers` which will not be used. OpenFF NAGL does not generate "
+                "conformers as part of assigning partial charges.",
+                UserWarning,
+                stacklevel=2,
+            )
+
+        if partial_charge_method == "_nagl_am1bccelf10":
```
Comment (j-wags) [inline openff/toolkit/utils/_nagl_wrapper.py:85]: Is the output of the nagl call really am1bccelf10 charges, or is it am1elf10 (and we need to apply the BCCs separately using something like a ChargeIncrementModelHandler)?

Comment (mattwthompson) [reply openff/toolkit/utils/_nagl_wrapper.py:85]: The output is AM1BCC ELF10 charges

**Comment by mattwthompson** [comment]
Tests are now only failing on charge differences. This wrapper can now progress without private API calls, but still lacks a model recommended for scientific use (and passing the expected behavior tests):

```python
In [1]: from openff.toolkit import Molecule

In [2]: from openff.nagl.nn._models import GNNModel
/Users/mattthompson/mambaforge/envs/openff-interchange-env/lib/python3.10/site-packages/lightning_utilities/core/imports.py:13: UserWarning: Module openff was already imported from None, but /Users/mattthompson/software/openff-interchange is being added to sys.path
  import pkg_resources

In [3]: from openff.nagl_models import validate_nagl_model_path, list_available_nagl_models

In [4]: model = GNNModel.load(validate_nagl_model_path(list_available_nagl_models()[0]))

In [5]: model.compute_property(Molecule.from_smiles("CCO"))
Out[5]:
tensor([[-0.1297],
        [-0.0904],
        [-0.1325],
        [ 0.0441],
        [ 0.0412],
        [ 0.1407],
        [ 0.0412],
        [ 0.0412],
        [ 0.0441]], grad_fn=<CatBackward0>)

In [6]: molecule = Molecule.from_smiles("CCO")

In [7]: molecule.assign_partial_charges(partial_charge_method="am1bccelf10")

In [8]: molecule.partial_charges
Out[8]:
array([-0.0971 ,  0.13143, -0.60134,  0.04476,  0.04476,  0.04476,
        0.01732,  0.01732,  0.39809]) <Unit('elementary_charge')>

In [9]: model.compute_property(Molecule.from_smiles("CCNCO"))
Out[9]:
tensor([[-0.0798],
        [ 0.1020],
        [-0.6875],
        [ 0.0965],
        [-0.1523],
        [ 0.3557],
        [ 0.0413],
        [ 0.0343],
        [ 0.1051],
        [ 0.0375],
        [ 0.0375],
        [ 0.0343],
        [ 0.0343],
        [ 0.0413]], grad_fn=<CatBackward0>)

In [10]: molecule = Molecule.from_smiles("CCNCO")

In [11]: molecule.assign_partial_charges(partial_charge_method="am1bccelf10")

In [12]: molecule.partial_charges
Out[12]:
array([-0.11312   ,  0.17023   , -0.83337999,  0.37266999, -0.59780998,
        0.04005   ,  0.04005   ,  0.04005   ,  0.03734999,  0.03734999,
        0.37525001,  0.01394   ,  0.01394   ,  0.40343001]) <Unit('elementary_charge')>

In [13]: list_available_nagl_models()
Out[13]: ['openff-gnn-am1bcc-0.0.1-alpha.1.pt']

**Review by j-wags — APPROVED** [review]
Remaining test failures seem to be unrelated Openmm/ParmEd stuff that I think I recall another PR fixing

**Comment by mattwthompson** [comment]
Those failures are new, probably related to ParmEd 4.1.0

**Pull request merged**